modules/login/jsx/loginIndex.js

```diff
@@ -0,0 +1,301 @@
+import PasswordExpired from './passwordExpiry';
+import RequestAccount from './requestAccount';
+import ResetPassword from './resetPassword';
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+import Loader from 'Loader';
+import Panel from 'Panel';
+
+/**
+ * Login form.
+ *
+ * @description form for login.
+ *
+ * @author Alizée Wickenheiser
+ * @version 1.0.0
+ *
+ */
+class Login extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      url: '',
+      study: {
+        logo: '',
+        title: '',
+        links: [],
+        description: '',
+      },
+      form: {
+        value: {
+          username: '',
+          password: '',
+        },
+        error: {
+          toggle: false,
+          message: '',
+        },
+      },
+      mode: 'login',
+      component: {
+        requestAccount: null,
+        expiredPassword: null,
+      },
+      isLoaded: false,
+    };
+    // Bind component instance to custom methods
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.fetchData = this.fetchData.bind(this);
+    this.setForm = this.setForm.bind(this);
+    this.setMode = this.setMode.bind(this);
+  }
+
+  /**
+   * Executes after component mounts.
+   */
+  componentDidMount() {
+    this.fetchData()
+      .then(() => this.setState({isLoaded: true}));
+  }
+
+  /**
+   * Retrieve data from the provided URL and save it in state.
+   *
+   * @return {object}
+   */
+  fetchData() {
+    return fetch(window.location.origin + '/login/Authentication',
+      {credentials: 'same-origin'}
+    )
+      .then((resp) => resp.json())
+      .then((json) => {
+        const state = Object.assign({}, this.state);
+        // login setup.
+        state.study.description = json.login.description;
+        state.study.title = json.login.title;
+        state.study.logo = window.location.origin
+          + '/' + json.login.logo;
+        // request account setup.
+        json.requestAccount.site[''] = 'Choose your site:';
+        state.component.requestAccount = json.requestAccount;
+        state.isLoaded = true;
+        this.setState(state);
+      }).catch((error) => {
+        this.setState({error: true});
+        console.error(error);
+      });
+  }
+
+  /**
+   * Set the form data based on state values of child elements/components
+   *
+   * @param {string} formElement - name of the selected element
+   * @param {string} value - selected value for corresponding form element
+   */
+  setForm(formElement, value) {
+    const state = Object.assign({}, this.state);
```

Comment:
I'm pretty sure we should be getting in the habit of using `JSON.parse(JSON.stringify(this.state))` since `Object.assign()` only creates a shallow copy. This also applies to lines 

https://stackoverflow.com/questions/41847581/what-is-the-difference-between-object-assign-and-json-parsejson-stringifyobj/41848531

According to this benchmark test, it also appears to be the fastest:
https://www.measurethat.net/Benchmarks/Show/2471/2/objectassign-vs-json-stringparse

Reply:
@HenriRabalais that's cool! How did you figure that out? Btw, I'm surprised there is nothing better than the `JSON.parse(JSON.stringify({..})` because I've seen or been told (whichever I cannot remember) that solution is a "hacky" way of making a new object. But that was years ago and maybe it's no longer considered hacky. I myself don't usually care if something is considered hacky by others if it works well! 😅

Reply:
@maltheism in 05 i still thought I was going to be an astronaut

Reply:
@maltheism I had been using it ever since Armin had recommended it as a proper way of cloning an object, and now every time I look into it when I doubt myself I see that it actually seems like the more efficient and robust solution!